src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -89,162 +99,179 @@ public class IndexDescriptor
-        public final IndexComponent component;
-
-        public AttachedIndexComponent(IndexComponent component, IndexContext context)
-        {
-            this.component = component;
-            this.context = context;
-        }
+        this.descriptor = descriptor;
+        this.partitioner = partitioner;
+        this.clusteringComparator = clusteringComparator;
 
-        @Override
-        public int hashCode()
-        {
-            return Objects.hash(context, component);
-        }
+        // Populating the per-sstable components. Not that this needs to happen first to have the proper version below.
+        populatePerSSTableComponents(this);
+        assert groups.containsKey(null);
 
-        @Override
-        public boolean equals(Object o)
-        {
-            if (this == o) return true;
-            if (o == null || getClass() != o.getClass()) return false;
-            AttachedIndexComponent that = (AttachedIndexComponent) o;
-            return Objects.equals(context, that.context) && component == that.component;
-        }
+        this.primaryKeyFactory = PrimaryKey.factory(clusteringComparator, getVersion().onDiskFormat().indexFeatureSet());
     }
 
-    private IndexDescriptor(Version version, Descriptor descriptor, IPartitioner partitioner, ClusteringComparator clusteringComparator)
+    private static void populatePerSSTableComponents(IndexDescriptor descriptor)
     {
-        this.descriptor = descriptor;
-        this.partitioner = partitioner;
-        this.clusteringComparator = clusteringComparator;
-        this.primaryKeyFactory = PrimaryKey.factory(clusteringComparator, version.onDiskFormat().indexFeatureSet());
-
-        versions.put(null, version);
-        components.put(null, Sets.newHashSet());
+        populateComponents(descriptor, null);
     }
 
-    public static IndexDescriptor createNew(Descriptor descriptor, IPartitioner partitioner, ClusteringComparator clusteringComparator)
+    private static void maybePopulateComponents(IndexDescriptor descriptor, IndexContext context)
     {
-        return new IndexDescriptor(Version.latest(), descriptor, partitioner, clusteringComparator);
+        if (!descriptor.groups.containsKey(context))
+            populateComponents(descriptor, context);
+
+        assert descriptor.groups.containsKey(context);
     }
 
-    public static IndexDescriptor createFrom(SSTableReader sstable)
+    private static void populateComponents(IndexDescriptor descriptor, @Nullable IndexContext context)
     {
-        // see if we have a completion component on disk, and if so use that version
+        // We first collect all the version/generation for which we have files on disk.
+        String indexName = context == null ? null : context.getIndexName();
+        Map<Version, IntObjectMap<Set<IndexComponent>>> candidates = Maps.newHashMap();
+
+        PathUtils.forEach(descriptor.descriptor.directory.toPath(), path -> {
```

Comment:
> How about checking the completion file (instead of listing the entire table directory) starting from generation 0 until no files are found for the generation?

I though about it, but there is 2 reasons I'm not sure this really work:
1. this assumes we do keep the files for every generation "forever" (forever in the sense of "as long as the sstable exists" which admittedly is not forever forever). And while this patch does not implement any deletion of old generation, I'd worry about not being able to at least write some async "cleanup" task for it later if we need to. I guess that cleanup task could delete all the generation file but the completion markers, but this still feel fragile, And cndb aside, I think this change is useful OSS-wise too, but relying on all the previous generation markers to be present for things to work feels particularly gratuitous fragility there: we can of course ignore that if we need to, but it's probably simpler if the implementation can work in both cases well enough.
2. the current patch actually delete completion markers when it detect a corruption to ensure that group of components does not get used, but still does not overwrite the generation in a rebuild. The idea is that even if we detect corruption in one component, there may still be other (non-corrupted) component used successfully and overridding files that are in used could still give fairly unpredictable results (which ultimately could well be worst than the corruption).

Overall, the idea here is to avoid overriding files that may still be in use concurrently from rebuilds, but those files still become unused eventually once the rebuild propagate, so I'm wary of the fragility of relying of those file sticking around no matter what.

> CNDB had some performance issues around listing files

Why is this? Isn't it a problem only if we do check remote storage?

Here, I think we'd be happy for the listing to only return what is either 1) local and 2) in the `RemoteFileCacheAttributes`, and it feel those should be fast (?). If so, we can maybe pass the sstable directory through some `StorageProvider` method that, for cndb, would decorate it with some open options that would indicate that we're happy to only use local information in the listing. Does that make sense, or am I missing the problem with listing?

Reply:
Thinking about this some more, I wonder if the simplest option isn't to add a new component (per "set of components") that stores the 'last written generation'. That component would obviously be special cases to be written without generation, so we could always read it without scanning. And if that file doesn't exists, we'd assume backward compat and generation 0 (in fact, we could dispense from writing the file at all if it's gen 0).

Reply:
> Why is this? Isn't it a problem only if we do check remote storage?
> listing to only return what is either 1) local and 2) in the RemoteFileCacheAttributes, and it feel those should be fast (?).

That's what `RemoteFileSystemProvider#newDirectoryStream` does now, both 1 and 2. I think `RemoteFileAttributesCache` is not optimized.

We can go with the current implementation and optimize later if it becomes an issue. `IndexDescriptor#create` is mainly used by writer when opening sstables and compactor. Taking a couple of milliseconds on a table with a thousand sstables should be fine.

Reply:
> Thinking about this some more, I wonder if the simplest option isn't to add a new component (per "set of components") that stores the 'last written generation'. That component would obviously be special cases to be written without generation, so we could always read it without scanning. And if that file doesn't exists, we'd assume backward compat and generation 0 (in fact, we could dispense from writing the file at all if it's gen 0).

that's TOC?